libraries/c_sdk/standard/mqtt/src/iot_mqtt_managed_function_wrapper.c

```diff
@@ -69,6 +69,177 @@ IotMqttError_t _IotMqtt_managedDisconnect( IotMqttConnection_t mqttConnection )
         /* Converting the status code. */
         status = convertReturnCode( managedMqttStatus );
     }
+    else
+    {
+        IotLogError( "(MQTT connection %p) MQTT Context is not set for this MQTT Connection.",
+                     mqttConnection );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+IotMqttError_t _IotMqtt_managedSubscribe( IotMqttConnection_t mqttConnection,
+                                          _mqttOperation_t * pSubscriptionOperation,
+                                          const IotMqttSubscription_t * pSubscriptionList,
+                                          size_t subscriptionCount )
+{
+    int8_t contextIndex = -1;
+    IotMqttError_t status = IOT_MQTT_BAD_PARAMETER;
+    /* Initializing MQTT Status. */
+    MQTTStatus_t managedMqttStatus = MQTTBadParameter;
+    uint16_t packetId = 0;
+    size_t i = 0;
+
+    IotMqtt_Assert( mqttConnection != NULL );
+    IotMqtt_Assert( pSubscriptionOperation != NULL );
+    IotMqtt_Assert( pSubscriptionList != NULL );
+
+    MQTTSubscribeInfo_t * subscriptionList = IotMqtt_MallocMessage( sizeof( MQTTSubscribeInfo_t ) * subscriptionCount );
+
+    /* Getting MQTT Context for the specified MQTT Connection. */
+    contextIndex = _IotMqtt_getContextIndexFromConnection( mqttConnection );
+
+    if( contextIndex >= 0 )
+    {
+        /* Generating the packet id for SUBSCRIBE packet. */
+        packetId = MQTT_GetPacketId( &( connToContext[ contextIndex ].context ) );
+        pSubscriptionOperation->u.operation.packetIdentifier = packetId;
+
+        /* Populating the subscription list to be used by MQTT LTS API. */
+        for( i = 0; i < subscriptionCount; i++ )
+        {
+            subscriptionList[ i ].qos = ( MQTTQoS_t ) ( pSubscriptionList + i )->qos;
+            subscriptionList[ i ].pTopicFilter = ( pSubscriptionList + i )->pTopicFilter;
+            subscriptionList[ i ].topicFilterLength = ( pSubscriptionList + i )->topicFilterLength;
+        }
+
+        IotMutex_Lock( &( connToContext[ contextIndex ].contextMutex ) );
+
+        /* Calling MQTT LTS API for sending the SUBSCRIBE packet on the network. */
+        managedMqttStatus = MQTT_Subscribe( &( connToContext[ contextIndex ].context ), subscriptionList, subscriptionCount, packetId );
```

Comment:
Why does this operation need to be surrounded by a mutex?  Likewise for the following MQTT_*() in the next few functions.

Reply:
All the MQTT_*() are surrounded by mutex so as to avoid corruption of buffer as the buffer is shared resource and all the managed API calls will be using it.